pubsub/psutil/psutil_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2018 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package psutil_test
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/mempubsub"
+	"gocloud.dev/pubsub/psutil"
+
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub/driver"
+)
+
+func TestReceiveConcurrently(t *testing.T) {
+	t.Run("can be cancelled", func(t *testing.T) {
+		top := mempubsub.NewTopic()
+		sub := mempubsub.NewSubscription(top, time.Second)
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel()
+		err := psutil.ReceiveConcurrently(ctx, sub, 10, func(ctx context.Context, m *pubsub.Message) error {
+			return nil // ack
+		})
+		if err == nil {
+			t.Fatal("got nil, want error")
+		}
+		if !strings.Contains(err.Error(), "cancel") {
```

Comment:
Better to check if `err == context.Canceled`.

Reply:
I wanted to do that, but that means I don't get to add a little bit of context using `fmt.Errorf`. Then it becomes hard to figure out where the error came from.